src/components/ImageUpload/ImageUploadProvider.tsx

```diff
@@ -0,0 +1,273 @@
+import { PromiseSettledResult } from '../../types';
+import { SnackbarDismissAction, SnackbarProps, useSnackbar } from '../Snackbar';
+import { useMaxUploadSize } from './useMaxUploadSize';
+
+/**
+ * Maximum allowed resolution in pixels.
+ */
+const maxResolutionPixels = 32_000_000;
+
+type UploadErrors = {
+  /** Name of files that failed due to size restrictions */
+  size: string[];
+  /** Name of files that failed due to resolution restrictions */
+  resolution: string[];
+  /** Name of files that failed for other reasons */
+  other: string[];
+};
+
+type ImageUploadResult = {
+  fileName: string;
+  size: Size;
+  /**
+   * MXC URI pointing to the uploaded image.
+   * {@link https://spec.matrix.org/v1.9/client-server-api/#matrix-content-mxc-uris}
+   */
+  mxc: string;
+};
+
+export type ImageUploadContextValue = {
+  /**
+   * Handle a file drop from react-dropzone.
+   * Show a loading snackbar as long as uploads are in progress.
+   * Once all uploads are finished show an error snackbar for errors, if any.
+   * Can be called multiple times while other uploads are still running.
+   *
+   * @param files - Files to be uploaded from react-dropzone
+   * @param rejectedFiles - Files rejected by react-dropzone
+   * @returns Results for the uploaded files
+   */
+  handleDrop: (
+    files: File[],
+    rejectedFiles: FileRejection[],
+  ) => Promise<PromiseSettledResult<ImageUploadResult>[]>;
+  /**
+   * Maximum upload size allowed by the server in bytes.
+   */
+  maxUploadSizeBytes: number;
+};
+
+class ResolutionTooHighError extends Error {
+  constructor() {
+    super('Resolution too high');
+  }
+}
+
+export const ImageUploadContext = createContext<
+  ImageUploadContextValue | undefined
+>(undefined);
+
+export function ImageUploadProvider({ children }: PropsWithChildren<{}>) {
+  const { t } = useTranslation();
+  const { clearSnackbar, showSnackbar } = useSnackbar();
+  const widgetApi = useWidgetApi();
+  /** Holds all files that are to be uploaded */
+  const uploadQueue = useRef(new Set<File>());
+  const uploadErrors = useRef<UploadErrors>({
+    size: [],
+    resolution: [],
+    other: [],
+  });
+  const { maxUploadSizeBytes: serverMaxUploadSizeBytes } = useMaxUploadSize();
+  // limit image uploads to 25 MiB
+  const maxUploadSizeBytes = Math.min(serverMaxUploadSizeBytes, 26214400);
+
+  const handleRejectedFile = useCallback((rejectedFile: FileRejection) => {
+    if (
+      rejectedFile.errors.some((error) => error.code === ErrorCode.FileTooLarge)
+    ) {
+      uploadErrors.current.size.push(rejectedFile.file.name);
+    } else {
+      uploadErrors.current.other.push(rejectedFile.file.name);
+    }
+  }, []);
+
+  const uploadImage = useCallback(
+    async (file: File) => {
+      uploadQueue.current.add(file);
+
+      try {
+        const imageData = await file.arrayBuffer();
+        const size = await determineImageSize(imageData);
+
+        if (size.width * size.height > maxResolutionPixels) {
```

Comment:
no 8k image support? ;)

Reply:
Nope. Unfortunately browsers are not ready, yet :snail: